File: frontend_ecommerce_manager_mobile-main/lib/services/customer_service_api.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';
import '../models/conversation.dart';
import '../services/auth_service.dart';
import '../utils/app_config.dart';

class CustomerServiceApiService {
  static String get baseUrl => '${AppConfig.apiBaseUrl}/customer-service';

  static Future<Map<String, String>> _getAuthHeaders() async {
    final headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    };

    final authService = AuthService();
    String? token;

    if (authService.currentUser != null &&
        authService.currentUser!.authToken.isNotEmpty) {
      token = authService.currentUser!.authToken;
    } else {
      final prefs = await SharedPreferences.getInstance();
      final currentEmail = prefs.getString('current_user_email');
      if (currentEmail != null) {
        token = await authService.getStoredAuthToken(currentEmail);
      }
    }

    if (token != null && token.isNotEmpty) {
      headers['auth_token'] = token;
    }

    return headers;
  }

  static Future<Map<String, dynamic>> getConversations(
    String shopId, {
    int pageSize = 20,
    String pageToken = '',
    String locale = 'id-ID',
  }) async {
    try {
      final queryParams = {
        'page_size': pageSize.toString(),
        'locale': locale,
        if (pageToken.isNotEmpty) 'page_token': pageToken,
      };

      final uri = Uri.parse('$baseUrl/$shopId/conversations')
          .replace(queryParameters: queryParams);

      final response = await http.get(uri, headers: await _getAuthHeaders());
      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        final conversations = (data['data']['conversations'] as List)
            .map((json) => Conversation.fromJson(json))
            .toList();

        return {
          'success': true,
          'conversations': conversations,
          'nextPageToken': data['data']['nextPageToken'],
          'shop': data['data']['shop'],
        };
      } else {
        throw Exception(data['message'] ?? 'Failed to get conversations');
      }
    } catch (e) {
      print('Error getting conversations: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> getMessages(
    String shopId,
    String conversationId, {
    int pageSize = 10,
    String pageToken = '',
    String locale = 'id-ID',
  }) async {
    try {
      final queryParams = {
        'page_size': pageSize.toString(),
        'locale': locale,
        if (pageToken.isNotEmpty) 'page_token': pageToken,
      };

      final uri =
          Uri.parse('$baseUrl/$shopId/conversations/$conversationId/messages')
              .replace(queryParameters: queryParams);

      final response = await http.get(uri, headers: await _getAuthHeaders());
      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        final messages = (data['data']['messages'] as List)
            .map((json) => Message.fromJson(json))
            .toList();

        return {
          'success': true,
          'messages': messages,
          'nextPageToken': data['data']['nextPageToken'],
        };
      } else {
        throw Exception(data['message'] ?? 'Failed to get messages');
      }
    } catch (e) {
      print('Error getting messages: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> createConversation(
    String shopId,
    String buyerUserId,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/$shopId/conversations'),
        headers: await _getAuthHeaders(),
        body: jsonEncode({
          'buyer_user_id': buyerUserId,
        }),
      );

      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        return {
          'success': true,
          'conversationId': data['data']['conversationId'],
        };
      } else {
        throw Exception(data['message'] ?? 'Failed to create conversation');
      }
    } catch (e) {
      print('Error creating conversation: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> sendMessage(
    String shopId,
    String conversationId,
    String type,
    String content,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/$shopId/conversations/$conversationId/messages'),
        headers: await _getAuthHeaders(),
        body: jsonEncode({
          'type': type,
          'content': content,
        }),
      );

      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        return {
          'success': true,
          'messageId': data['data']['messageId'],
        };
      } else {
        throw Exception(data['message'] ?? 'Failed to send message');
      }
    } catch (e) {
      print('Error sending message: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> readMessages(
    String shopId,
    String conversationId,
  ) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/$shopId/conversations/$conversationId/read'),
        headers: await _getAuthHeaders(),
      );

      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        return {'success': true};
      } else {
        throw Exception(data['message'] ?? 'Failed to mark messages as read');
      }
    } catch (e) {
      print('Error marking messages as read: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> uploadImage(
    String shopId,
    File imageFile,
  ) async {
    try {
      final request = http.MultipartRequest(
        'POST',
        Uri.parse('$baseUrl/$shopId/images/upload'),
      );

      final headers = await _getAuthHeaders();
      headers.remove('Content-Type');
      request.headers.addAll(headers);

      request.files.add(
        await http.MultipartFile.fromPath('data', imageFile.path),
      );

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      final data = jsonDecode(response.body);

      if (response.statusCode == 200 && data['success']) {
        return {
          'success': true,
          'url': data['data']['url'],
          'width': data['data']['width'],
          'height': data['data']['height'],
        };
      } else {
        throw Exception(data['message'] ?? 'Failed to upload image');
      }
    } catch (e) {
      print('Error uploading image: $e');
      return {'success': false, 'error': e.toString()};
    }
  }

  static Future<Map<String, dynamic>> sendTextMessage(
    String shopId,
    String conversationId,
    String text,
  ) async {
    final content = jsonEncode({'content': text});
    return sendMessage(shopId, conversationId, 'TEXT', content);
  }

  static Future<Map<String, dynamic>> sendProductCard(
    String shopId,
    String conversationId,
    String productId,
  ) async {
    final content = jsonEncode({'product_id': productId});
    return sendMessage(shopId, conversationId, 'PRODUCT_CARD', content);
  }

  static Future<Map<String, dynamic>> sendOrderCard(
    String shopId,
    String conversationId,
    String orderId,
  ) async {
    final content = jsonEncode({'order_id': orderId});
    return sendMessage(shopId, conversationId, 'ORDER_CARD', content);
  }
}

File: backend_ecommerce_manager_mobile-main/src/services/tiktok/customerServiceAPI.js
// backend_ecommerce_manager_mobile/src/services/customerServiceAPI.js
const { generateTikTokSignature } = require("../../utils/tiktokSignature");
const config = require("../../config/env");

class CustomerServiceAPI {
  constructor() {
    this.baseUrl = "https://open-api.tiktokglobalshop.com";
    this.apiVersion = "202309";
  }

  // Helper untuk generate common parameters
  generateCommonParams() {
    const timestamp = Math.floor(Date.now() / 1000);
    return {
      app_key: config.tiktok.appKey,
      timestamp: timestamp.toString(),
    };
  }

  // Helper untuk create URL with signature
  createSignedUrl(path, queryParams, accessToken) {
    const signature = generateTikTokSignature(
      path,
      queryParams,
      "",
      config.tiktok.appSecret,
    );
    queryParams.sign = signature;

    const queryString = new URLSearchParams(queryParams).toString();
    return `${this.baseUrl}${path}?${queryString}`;
  }

  // 1. Get Conversations
  async getConversations(
    accessToken,
    shopCipher,
    pageSize = 20,
    pageToken = "",
    locale = "id-ID",
  ) {
    try {
      const path = `/customer_service/${this.apiVersion}/conversations`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
        page_size: pageSize.toString(),
        locale: locale,
      };

      if (pageToken) {
        queryParams.page_token = pageToken;
      }

      const url = this.createSignedUrl(path, queryParams);

      const response = await fetch(url, {
        method: "GET",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "x-tts-access-token": accessToken,
        },
      });

      const data = await response.json();
      console.log(" Get Conversations API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in getConversations:", error);
      throw error;
    }
  }

  // 2. Get Conversation Messages
  async getConversationMessages(
    accessToken,
    shopCipher,
    conversationId,
    pageSize = 10,
    pageToken = "",
    locale = "id-ID",
  ) {
    try {
      const path = `/customer_service/${this.apiVersion}/conversations/${conversationId}/messages`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
        page_size: pageSize.toString(),
        locale: locale,
        sort_order: "DESC",
        sort_field: "create_time",
      };

      if (pageToken) {
        queryParams.page_token = pageToken;
      }

      const url = this.createSignedUrl(path, queryParams);

      const response = await fetch(url, {
        method: "GET",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "x-tts-access-token": accessToken,
        },
      });

      const data = await response.json();
      console.log(" Get Messages API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in getConversationMessages:", error);
      throw error;
    }
  }

  // 3. Create Conversation
  async createConversation(accessToken, shopCipher, buyerUserId) {
    try {
      const path = `/customer_service/${this.apiVersion}/conversations`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
      };

      const body = JSON.stringify({
        buyer_user_id: buyerUserId,
      });

      const url = this.createSignedUrl(path, queryParams, body);

      const response = await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "x-tts-access-token": accessToken,
        },
        body: body,
      });

      const data = await response.json();
      console.log(" Create Conversation API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in createConversation:", error);
      throw error;
    }
  }

  // 4. Send Message
  async sendMessage(
    accessToken,
    shopCipher,
    conversationId,
    messageType,
    content,
  ) {
    try {
      const path = `/customer_service/${this.apiVersion}/conversations/${conversationId}/messages`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
      };

      const body = JSON.stringify({
        type: messageType,
        content: content,
      });

      const url = this.createSignedUrl(path, queryParams, body);

      const response = await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "x-tts-access-token": accessToken,
        },
        body: body,
      });

      const data = await response.json();
      console.log(" Send Message API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in sendMessage:", error);
      throw error;
    }
  }

  // 5. Read Messages
  async readMessages(accessToken, shopCipher, conversationId) {
    try {
      const path = `/customer_service/${this.apiVersion}/conversations/${conversationId}/messages/read`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
      };

      const url = this.createSignedUrl(path, queryParams);

      const response = await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "Content-Type": "application/json",
          "x-tts-access-token": accessToken,
        },
      });

      const data = await response.json();
      console.log("️ Read Messages API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in readMessages:", error);
      throw error;
    }
  }

  // 6. Upload Image
  async uploadImage(accessToken, shopCipher, imageFile) {
    try {
      const path = `/customer_service/${this.apiVersion}/images/upload`;
      const queryParams = {
        ...this.generateCommonParams(),
        shop_cipher: shopCipher,
      };

      const url = this.createSignedUrl(path, queryParams);

      const formData = new FormData();
      formData.append("data", imageFile);

      const response = await fetch(url, {
        method: "POST",
        headers: {
          Accept: "application/json",
          "x-tts-access-token": accessToken,
        },
        body: formData,
      });

      const data = await response.json();
      console.log("️ Upload Image API Response:", data.code);
      return data;
    } catch (error) {
      console.error(" Error in uploadImage:", error);
      throw error;
    }
  }
}

module.exports = new CustomerServiceAPI();
